src/ripple/nodestore/DatabaseShard.h

```diff
@@ -0,0 +1,176 @@
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_DATABASESHARD_H_INCLUDED
+#define RIPPLE_NODESTORE_DATABASESHARD_H_INCLUDED
+
+#include <ripple/nodestore/Database.h>
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/nodestore/Types.h>
+
+#include <boost/optional.hpp>
+
+#include <memory>
+
+namespace ripple {
+namespace NodeStore {
+
+/** A collection of historical shards
+*/
+class DatabaseShard : public Database
+{
+public:
+    /** Construct a shard store
+
+        @param name The Stoppable name for this Database
+        @param parent The parent Stoppable
+        @param scheduler The scheduler to use for performing asynchronous tasks
+        @param readThreads The number of async read threads to create
+        @param config The configuration for the database
+        @param journal Destination for logging output
+    */
+    DatabaseShard(std::string const& name, Stoppable& parent,
+        Scheduler& scheduler, int readThreads,
+        Section const& config, beast::Journal journal)
+        : Database(name, parent, scheduler, readThreads, journal)
+    {
+        get_if_exists<std::uint32_t>(config, "ledgers_per_shard", lps_);
+    }
+
+    /** Initialize the database
+
+        @return `true` if the database initialized without error
+    */
+    virtual
+    bool
+    init() = 0;
+
+    /** Prepare to store a new ledger in the shard
+
+        @param validLedgerSeq the index of the maximum valid ledgers
+        @return if a ledger should be fetched and stored, then returns the ledger
+                index of the ledger to request. Otherwise returns boost::none.
+                Some reasons this may return boost::none are: this database does
+                not store shards, all shards are are stored and full, max allowed
+                disk space would be exceeded, or a ledger was recently requested
+                and not enough time has passed between requests.
+        @implNote adds a new writable shard if necessary
+    */
+    virtual
+    boost::optional<std::uint32_t>
+    prepare(std::uint32_t validLedgerSeq) = 0;
+
+    /** Fetch a ledger from the shard store
+
+        @param hash The key of the ledger to retrieve
+        @param seq The sequence of the ledger
+        @return The ledger if found, nullptr otherwise
+    */
+    virtual
+    std::shared_ptr<Ledger>
+    fetchLedger(uint256 const& hash, std::uint32_t seq) = 0;
+
+    /** Notifies the database that the given ledger has been
+        fully acquired and stored.
+
+        @param ledger The stored ledger to be marked as complete
+    */
+    virtual
+    void
+    setStored(std::shared_ptr<Ledger const> const& ledger) = 0;
+
+    /** Query if a ledger with the given sequence is stored
+
+        @param seq The ledger sequence to check if stored
+        @return `true` if the ledger is stored
+    */
+    virtual
+    bool
+    hasLedger(std::uint32_t seq) = 0;
```

Comment:
No change needed, just food for thought. Should we prefer generic names like `contains` over something like `hasLedger`?

Reply:
Definitely.